File: src/common/Logger.ts
import { invoke } from "@tauri-apps/api/core";

class Logger {
  // 分隔线
  private readonly LINE = "==================================================";
  private readonly SMALL_LINE =
    "--------------------------------------------------";

  // 发送给 Rust 打印
  private async print(level: string, title: string, content: string) {
    const log = `
${this.LINE}
[${level}] ${title}
${this.SMALL_LINE}
${content}
${this.LINE}
`;
    await invoke("js_log", { message: log.trim() });
  }

  // 信息日志
  async info(title: string, ...args: any[]) {
    const msg = this.formatArgs(args);
    await this.print("INFO", title, msg);
  }

  // 错误日志
  async error(title: string, ...args: any[]) {
    const msg = this.formatArgs(args);
    await this.print("ERROR", title, msg);
  }

  // 成功日志
  async success(title: string, ...args: any[]) {
    const msg = this.formatArgs(args);
    await this.print("SUCCESS", title, msg);
  }

  // 把所有参数转成可打印字符串
  private formatArgs(args: any[]): string {
    return args
      .map((item) => {
        if (typeof item === "object") {
          try {
            return JSON.stringify(item, null, 2);
          } catch {
            return String(item);
          }
        }
        return String(item);
      })
      .join("\n");
  }
}

export default new Logger();

File: src-tauri/src/logger.rs
/// Rust Logger 模块
/// 参考 src/common/Logger.ts 设计
pub struct Logger;

impl Logger {
    const LINE: &'static str = "==================================================";
    const SMALL_LINE: &'static str = "--------------------------------------------------";

    /// 格式化参数为字符串
    #[allow(dead_code)]
    fn format_args(args: &[impl std::fmt::Debug]) -> String {
        if args.is_empty() {
            return String::new();
        }
        args.iter()
            .map(|item| format!("{:?}", item))
            .collect::<Vec<_>>()
            .join("\n")
    }

    /// 打印格式化后的日志
    #[allow(dead_code)]
    fn print(level: &str, title: &str, content: &str) {
        let log = if content.is_empty() {
            format!("{}\n[{}] {}\n{}", Self::LINE, level, title, Self::LINE)
        } else {
            format!(
                "{}\n[{}] {}\n{}\n{}\n{}",
                Self::LINE,
                level,
                title,
                Self::SMALL_LINE,
                content,
                Self::LINE
            )
        };
        println!("{}", log);
    }

    /// 信息日志
    #[allow(dead_code)]
    pub fn info<T: std::fmt::Debug>(title: &str, args: &[T]) {
        let content = Self::format_args(args);
        Self::print("INFO", title, &content);
    }

    /// 错误日志
    #[allow(dead_code)]
    pub fn error<T: std::fmt::Debug>(title: &str, args: &[T]) {
        let content = Self::format_args(args);
        Self::print("ERROR", title, &content);
    }

    /// 成功日志
    #[allow(dead_code)]
    pub fn success<T: std::fmt::Debug>(title: &str, args: &[T]) {
        let content = Self::format_args(args);
        Self::print("SUCCESS", title, &content);
    }
}
